File: src/main/java/ru/nsu/ccfit/dmakogon/NodeState.java
package ru.nsu.ccfit.dmakogon;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;

import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import ru.nsu.ccfit.dmakogon.network.PeersList;
import ru.nsu.ccfit.dmakogon.operations.OperationsLog;

@RequiredArgsConstructor
public class NodeState {
  public static final int NOT_VOTED_ID = -1;

  public enum NodeType {
    LEADER, CANDIDATE, FOLLOWER
  }

  @Getter
  @Setter
  private NodeType type = NodeType.CANDIDATE;

  private final AtomicLong currentTerm = new AtomicLong(0);
  private final AtomicInteger votedFor = new AtomicInteger(0);
  private final AtomicInteger commitIndex = new AtomicInteger(0);
  private final AtomicInteger lastApplied = new AtomicInteger(0);
  private final PeersList peers;
  private final OperationsLog operationsLog;

  public PeersList getPeers() {
    return peers;
  }

  public int getQuorum() {
    return (peers.size() + 1) / 2 + 1;
  }

  public OperationsLog getOperationsLog() {
    return operationsLog;
  }

  public long getCurrentTerm() {
    return currentTerm.get();
  }

  public boolean setCurrentTermToGreater(long term) {
    return currentTerm.updateAndGet(x -> Math.max(x, term)) < term;
  }

  public Optional<Peer> findPeer(int id) {
    var peersWithId = peers.stream()
            .filter(peer -> peer.getId() == id)
            .toList();
    if (peersWithId.size() > 1)
      throw new IllegalStateException("Several peers with same ids");
    if (peersWithId.size() == 1)
      return Optional.of(peersWithId.get(0));
    return Optional.empty();
  }

  public void setCurrentTerm(long currentTerm) {
    this.currentTerm.set(currentTerm);
  }

  public long incrementCurrentTerm() {
    return this.currentTerm.incrementAndGet();
  }

  public Optional<Integer> getVotedFor() {
    var value = votedFor.get();
    if (value == NOT_VOTED_ID)
      return Optional.empty();
    return Optional.of(value);
  }

  public void setVotedFor(int votedFor) {
    this.votedFor.set(votedFor);
  }

  public int getCommitIndex() {
    return commitIndex.get();
  }

  public int getLastApplied() {
    return lastApplied.get();
  }

  public int incrementLastApplied() {
    return lastApplied.incrementAndGet();
  }

  public void setCommitIndex(int commitIndex) {
    this.commitIndex.set(commitIndex);
  }
}

File: sync-hashmap-python/node_state.py
from enum import Enum
from typing import Optional
from atomic import Atomic, AtomicInt
from operations_log import OperationsLog

from peer import PeerId, PeersList


class NodeType(Enum):
    LEADER = 1
    CANDIDATE = 2
    FOLLOWER = 3


class NodeState:
    NOT_VOTED_ID = PeerId(-1)

    type: NodeType
    currentTerm: AtomicInt
    votedFor: Atomic[PeerId]
    commitIndex: AtomicInt
    lastApplied: AtomicInt
    peersList: PeersList
    operationsLog: OperationsLog

    def getQuorum(self) -> int:
        return int((len(self.peersList) + 1) / 2 + 1)

    def getCurrentTerm(self) -> int:
        return self.currentTerm.get()

    def setCurrentTermToGreater(self, term: int) -> bool:
        return self.currentTerm.getAndUpdate(lambda x: max(x, term)) < term

    def incrementCurrentTerm(self) -> int:
        return self.currentTerm.incrementAndGet()

    def getVotedFor(self) -> Optional[PeerId]:
        value = self.votedFor.get()
        if value.isValid():
            return value
        return None

    def setVotedFor(self, votedFor: PeerId):
        self.votedFor.set(votedFor)

    def getCommitIndex(self):
        return self.commitIndex.get()

    def getLastApplied(self):
        return self.lastApplied.get()

    def incrementLastApplied(self):
        return self.lastApplied.incrementAndGet()

    def setCommitIndex(self, commitIndex: int):
        self.commitIndex.set(commitIndex)
